File: backend/app/schemas/crowd.py
from pydantic import BaseModel, ConfigDict, Field
from datetime import datetime
import uuid


class ZoneDensityResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    zone_id: uuid.UUID
    venue_id: uuid.UUID
    zone_name: str
    density: float = Field(ge=0.0, le=1.0)
    level: str  # LOW|MEDIUM|HIGH|CRITICAL
    count: int
    is_open: bool
    updated_at: datetime | None


class UpdateDensityRequest(BaseModel):
    density: float = Field(ge=0.0, le=1.0)
    count: int = Field(ge=0)
    venue_id: uuid.UUID


class VenueDensitySummary(BaseModel):
    venue_id: uuid.UUID
    total_zones: int
    critical_zones: int
    high_zones: int
    medium_zones: int
    low_zones: int
    overall_level: str
    zones: list[ZoneDensityResponse]


class CongestionPrediction(BaseModel):
    zone_id: uuid.UUID
    zone_name: str
    predicted_level_15min: str
    predicted_level_30min: str
    confidence: float = Field(ge=0.0, le=1.0)
    alert: bool
    alert_message: str | None = None

File: frontend/src/types/crowd.ts
export type DensityLevel = 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';

export interface ZoneDensity {
  readonly id: string;
  readonly zoneId: string;
  readonly venueId: string;
  readonly venue_id?: string;
  readonly name: string;
  readonly zoneName: string;
  readonly zone_name?: string;
  readonly zone_type?: string | null;
  readonly density: number;
  readonly current_density?: number;
  readonly level: DensityLevel;
  readonly density_level?: DensityLevel;
  readonly count: number;
  readonly isOpen: boolean;
  readonly is_open?: boolean;
  readonly updatedAt: string | null;
  readonly updated_at?: string | null;
  readonly latCenter?: number;
  readonly lat_center?: number;
  readonly lngCenter?: number;
  readonly lng_center?: number;
  readonly polygon?: Array<{ lat: number; lng: number }> | null;
  readonly center_point?: { latitude: number; longitude: number } | null;
}

export interface VenueDensitySummary {
  readonly venueId: string;
  readonly totalZones: number;
  readonly criticalZones: number;
  readonly highZones: number;
  readonly mediumZones: number;
  readonly lowZones: number;
  readonly overallLevel: DensityLevel;
  readonly zones: ZoneDensity[];
}

export interface CongestionPrediction {
  readonly zoneId: string;
  readonly zone_id?: string;
  readonly zoneName: string;
  readonly zone_name?: string;
  readonly predictedLevel15min: DensityLevel;
  readonly predicted_level_15min?: DensityLevel;
  readonly predictedLevel30min: DensityLevel;
  readonly predicted_level_30min?: DensityLevel;
  readonly confidence: number;
  readonly alert: boolean;
  readonly alertMessage: string | null;
  readonly alert_message?: string | null;
}

export interface FirestoreZoneDensity {
  density: number;
  level: DensityLevel;
  count: number;
  updated_at: { toDate: () => Date } | null;
}
